File: modules/core/src/main/java/org/apache/ignite/cache/query/ContinuousQuery.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.ignite.cache.query;

import javax.cache.Cache;
import javax.cache.configuration.Factory;
import javax.cache.event.CacheEntryEventFilter;
import javax.cache.event.CacheEntryUpdatedListener;
import org.apache.ignite.IgniteCache;
import org.apache.ignite.cache.CacheEntryEventSerializableFilter;
import org.apache.ignite.cache.query.ContinuousQueryWithTransformer.EventListener;
import org.apache.ignite.configuration.IgniteConfiguration;
import org.apache.ignite.lang.IgniteAsyncCallback;

/**
 * API for configuring continuous cache queries.
 * <p>
 * Continuous queries allow registering a remote filter and a local listener
 * for cache updates. If an update event passes the filter, it will be sent to
 * the node that executed the query, and local listener will be notified.
 * <p>
 * Additionally, you can execute an initial query to get currently existing data.
 * Query can be of any type (SQL, TEXT or SCAN) and can be set via {@link #setInitialQuery(Query)}
 * method.
 * <p>
 * Query can be executed either on all nodes in topology using {@link IgniteCache#query(Query)}
 * method, or only on the local node, if {@link Query#setLocal(boolean)} parameter is set to {@code true}.
 * Note that if the query is distributed and a new node joins, it will get the remote
 * filter for the query during discovery process before it actually joins a topology,
 * so no updates will be missed.
 * <h1 class="header">Example</h1>
 * As an example, suppose we have a cache with {@code 'Person'} objects and we need
 * to query for all people with salary above 1000.
 * <p>
 * Here is the {@code Person} class:
 * <pre name="code" class="java">
 * public class Person {
 *     // Name.
 *     private String name;
 *
 *     // Salary.
 *     private double salary;
 *
 *     ...
 * }
 * </pre>
 * <p>
 * You can create and execute a continuous query like so:
 * <pre name="code" class="java">
 * // Create a new continuous query.
 * ContinuousQuery&lt;Long, Person&gt; qry = new ContinuousQuery&lt;&gt;();
 *
 * // Initial iteration query will return all people with salary above 1000.
 * qry.setInitialQuery(new ScanQuery&lt;&gt;((id, p) -> p.getSalary() &gt; 1000));
 *
 *
 * // Callback that is called locally when update notifications are received.
 * // It simply prints out information about all created or modified records.
 * qry.setLocalListener((evts) -> {
 *     for (CacheEntryEvent&lt;? extends Long, ? extends Person&gt; e : evts) {
 *         Person p = e.getValue();
 *
 *         System.out.println(p.getFirstName() + " " + p.getLastName() + "'s salary is " + p.getSalary());
 *     }
 * });
 *
 * // The continuous listener will be notified for people with salary above 1000.
 * qry.setRemoteFilter(evt -> evt.getValue().getSalary() &gt; 1000);
 *
 * // Execute the query and get a cursor that iterates through the initial data.
 * QueryCursor&lt;Cache.Entry&lt;Long, Person&gt;&gt; cur = cache.query(qry);
 * </pre>
 * This will execute query on all nodes that have the cache you are working with and
 * listener will start receiving notifications for cache updates.
 * <p>
 * To stop receiving updates call {@link QueryCursor#close()} method:
 * <pre name="code" class="java">
 * cur.close();
 * </pre>
 * Note that this works even if you didn't provide the initial query. Cursor will
 * be empty in this case, but it will still unregister listeners when {@link QueryCursor#close()}
 * is called.
 * <p>
 * {@link IgniteAsyncCallback} annotation is supported for {@link CacheEntryEventFilter}
 * (see {@link #setRemoteFilterFactory(Factory)}) and {@link CacheEntryUpdatedListener}
 * (see {@link #setLocalListener(CacheEntryUpdatedListener)}).
 * If a filter and/or listener are annotated with {@link IgniteAsyncCallback} then the annotated callback
 * is executed in an async callback pool (see {@link IgniteConfiguration#getAsyncCallbackPoolSize()})
 * and a notification order is kept the same as an update order for a given cache key.
 *
 * @see ContinuousQueryWithTransformer
 * @see IgniteAsyncCallback
 * @see IgniteConfiguration#getAsyncCallbackPoolSize()
 */
public final class ContinuousQuery<K, V> extends AbstractContinuousQuery<K, V> {
    /** */
    private static final long serialVersionUID = 0L;

    /** Local listener. */
    private CacheEntryUpdatedListener<K, V> locLsnr;

    /** Remote filter. */
    private CacheEntryEventSerializableFilter<K, V> rmtFilter;

    /**
     * Creates new continuous query.
     */
    public ContinuousQuery() {
        setPageSize(DFLT_PAGE_SIZE);
    }

    /** {@inheritDoc} */
    @Override public ContinuousQuery<K, V> setInitialQuery(Query<Cache.Entry<K, V>> initQry) {
        return (ContinuousQuery<K, V>)super.setInitialQuery(initQry);
    }

    /**
     * Sets a local callback. This callback is called only on local node when new updates are received.
     * <p>
     * The callback predicate accepts ID of the node from where updates are received and a collection
     * of the received entries. Note that for removed entries values will be {@code null}.
     * <p>
     * If the predicate returns {@code false}, query execution will be cancelled.
     * <p>
     * <b>WARNING:</b> all operations that involve any kind of JVM-local or distributed locking (e.g.,
     * synchronization or transactional cache operations), should be executed asynchronously without
     * blocking the thread that called the callback. Otherwise, you can get deadlocks.
     * <p>
     * If local listener are annotated with {@link IgniteAsyncCallback} then it is executed in an async callback pool
     * (see {@link IgniteConfiguration#getAsyncCallbackPoolSize()}) that allow to perform a cache operations.
     *
     * @param locLsnr Local callback.
     * @return {@code this} for chaining.
     * @see IgniteAsyncCallback
     * @see IgniteConfiguration#getAsyncCallbackPoolSize()
     * @see ContinuousQueryWithTransformer#setLocalListener(EventListener)
     */
    public ContinuousQuery<K, V> setLocalListener(CacheEntryUpdatedListener<K, V> locLsnr) {
        this.locLsnr = locLsnr;

        return this;
    }

    /**
     * @return Local listener.
     */
    public CacheEntryUpdatedListener<K, V> getLocalListener() {
        return locLsnr;
    }

    /**
     * Sets optional key-value filter. This filter is called before entry is sent to the master node.
     * <p>
     * <b>WARNING:</b> all operations that involve any kind of JVM-local or distributed locking
     * (e.g., synchronization or transactional cache operations), should be executed asynchronously
     * without blocking the thread that called the filter. Otherwise, you can get deadlocks.
     * <p>
     * If remote filter are annotated with {@link IgniteAsyncCallback} then it is executed in async callback
     * pool (see {@link IgniteConfiguration#getAsyncCallbackPoolSize()}) that allow to perform a cache operations.
     *
     * @param rmtFilter Key-value filter.
     * @return {@code this} for chaining.
     *
     * @deprecated Use {@link #setRemoteFilterFactory(Factory)} instead.
     * @see IgniteAsyncCallback
     * @see IgniteConfiguration#getAsyncCallbackPoolSize()
     */
    @Deprecated
    public ContinuousQuery<K, V> setRemoteFilter(CacheEntryEventSerializableFilter<K, V> rmtFilter) {
        this.rmtFilter = rmtFilter;

        return this;
    }

    /**
     * Gets remote filter.
     *
     * @return Remote filter.
     */
    public CacheEntryEventSerializableFilter<K, V> getRemoteFilter() {
        return rmtFilter;
    }

    /** {@inheritDoc} */
    @Override public ContinuousQuery<K, V> setTimeInterval(long timeInterval) {
        return (ContinuousQuery<K, V>)super.setTimeInterval(timeInterval);
    }

    /** {@inheritDoc} */
    @Override public ContinuousQuery<K, V> setAutoUnsubscribe(boolean autoUnsubscribe) {
        return (ContinuousQuery<K, V>)super.setAutoUnsubscribe(autoUnsubscribe);
    }

    /** {@inheritDoc} */
    @Override public ContinuousQuery<K, V> setPageSize(int pageSize) {
        return (ContinuousQuery<K, V>)super.setPageSize(pageSize);
    }

    /**
     * Sets whether this query should be executed on a local node only.
     *
     * Note: backup event queues are not kept for local continuous queries. It may lead to loss of notifications in case
     * of node failures. Use {@link ContinuousQuery#setRemoteFilterFactory(Factory)} to register cache event listeners
     * on all cache nodes, if delivery guarantee is required.
     *
     * @param loc Local flag.
     * @return {@code this} for chaining.
     */
    @Override public ContinuousQuery<K, V> setLocal(boolean loc) {
        return (ContinuousQuery<K, V>)super.setLocal(loc);
    }
}

File: modules/platforms/dotnet/Apache.Ignite.Core/Cache/Query/Continuous/ContinuousQuery.cs
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace Apache.Ignite.Core.Cache.Query.Continuous
{
    using System;
    using Apache.Ignite.Core.Cache.Event;

    /// <summary>
    /// Continuous query base class, see <see cref="ContinuousQuery{K, V}"/>.
    /// </summary>
    public abstract class ContinuousQuery
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="ContinuousQuery"/> class.
        /// </summary>
        protected internal ContinuousQuery()
        {
            // No-op.
        }

        /// <summary>
        /// Default buffer size.
        /// </summary>
        public const int DefaultBufferSize = 1;

        /// <summary>
        /// Default time interval.
        /// </summary>
        public static readonly TimeSpan DefaultTimeInterval = new TimeSpan(0);

        /// <summary>
        /// Default auto-unsubscribe flag value.
        /// </summary>
        public const bool DefaultAutoUnsubscribe = true;
    }

    /// <summary>
    /// API for configuring continuous cache queries.
    /// <para />
    /// Continuous queries allow to register a remote and a listener for cache update events.
    /// If an update event passes the filter, it will be sent to the node that executed the
    /// query and listener will be notified on that node.
    /// <para />
    /// Continuous query can either be executed on the whole topology or only on local node.
    /// <para />
    /// In case query is distributed and a new node joins, it will get the filter for the query
    /// during discovery process before it actually joins topology, so no updates will be missed.
    /// <para />
    /// To execute the query use method
    /// <see cref="ICache{K,V}.QueryContinuous(ContinuousQuery{K,V})"/>.
    /// </summary>
    public class ContinuousQuery<TK, TV> : ContinuousQuery
    {
        /// <summary>
        /// Constructor.
        /// </summary>
        /// <param name="lsnr">Listener.</param>
        public ContinuousQuery(ICacheEntryEventListener<TK, TV> lsnr) : this(lsnr, false)
        {
            // No-op.
        }

        /// <summary>
        /// Constructor.
        /// </summary>
        /// <param name="lsnr">Listener.</param>
        /// <param name="loc">Whether query should be executed locally.</param>
        public ContinuousQuery(ICacheEntryEventListener<TK, TV> lsnr, bool loc) : this(lsnr, null, loc)
        {
            // No-op.
        }

        /// <summary>
        /// Constructor.
        /// </summary>
        /// <param name="lsnr">Listener.</param>
        /// <param name="filter">Filter.</param>
        public ContinuousQuery(ICacheEntryEventListener<TK, TV> lsnr, ICacheEntryEventFilter<TK, TV> filter)
            : this(lsnr, filter, false)
        {
            // No-op.
        }

        /// <summary>
        /// Constructor.
        /// </summary>
        /// <param name="lsnr">Listener.</param>
        /// <param name="filter">Filter.</param>
        /// <param name="loc">Whether query should be executed locally.</param>
        public ContinuousQuery(ICacheEntryEventListener<TK, TV> lsnr, ICacheEntryEventFilter<TK, TV> filter, bool loc)
        {
            Listener = lsnr;
            Filter = filter;
            Local = loc;

            BufferSize = DefaultBufferSize;
            TimeInterval = DefaultTimeInterval;
            AutoUnsubscribe = DefaultAutoUnsubscribe;
        }

        /// <summary>
        /// Cache entry event listener. Invoked on the node where continuous query execution
        /// has been started.
        /// </summary>
        public ICacheEntryEventListener<TK, TV> Listener { get; set; }

        /// <summary>
        /// Optional cache entry filter. Invoked on a node where cache event occurred. If filter
        /// returns <c>false</c>, then cache entry event will not be sent to a node where
        /// continuous query has been started.
        /// <para />
        /// Must be either binary or serializable in case query is not local.
        /// </summary>
        public ICacheEntryEventFilter<TK, TV> Filter { get; set; }

        /// <summary>
        /// Buffer size. When a cache update happens, entry is first put into a buffer.
        /// Entries from buffer will be sent to the master node only if the buffer is
        /// full or time provided via <see cref="TimeInterval"/> is exceeded.
        /// <para />
        /// Defaults to <see cref="ContinuousQuery.DefaultBufferSize"/>
        /// </summary>
        public int BufferSize { get; set; }

        /// <summary>
        /// Time interval. When a cache update happens, entry is first put into a buffer.
        /// Entries from buffer will be sent to the master node only if the buffer is full
        /// (its size can be provided via <see cref="BufferSize"/> property) or time provided
        /// via this method is exceeded.
        /// <para />
        /// Defaults to <c>0</c> which means that time check is disabled and entries will be
        /// sent only when buffer is full.
        /// </summary>
        public TimeSpan TimeInterval { get; set; }

        /// <summary>
        /// Automatic unsubscribe flag. This flag indicates that query filters on remote nodes
        /// should be automatically unregistered if master node (node that initiated the query)
        /// leaves topology. If this flag is <c>false</c>, filters will be unregistered only
        /// when the query is cancelled from master node, and won't ever be unregistered if
        /// master node leaves grid.
        /// <para />
        /// Defaults to <c>true</c>.
        /// </summary>
        public bool AutoUnsubscribe { get; set; }

        /// <summary>
        /// Local flag. When set query will be executed only on local node, so only local
        /// entries will be returned as query result.
        /// <para />
        /// Defaults to <c>false</c>.
        /// </summary>
        public bool Local { get; set; }
        
        /// <summary>
        /// Gets or sets a value indicating whether to notify about <see cref="CacheEntryEventType.Expired"/> events.
        /// <para />
        /// If <c>true</c>, then the remote listener will get notifications about expired cache entries.
        /// Otherwise, only <see cref="CacheEntryEventType.Created"/>, <see cref="CacheEntryEventType.Updated"/>, and
        /// <see cref="CacheEntryEventType.Removed"/> events will be passed to the listener.
        /// <para />
        /// Defaults to <c>false</c>.
        /// </summary>
        public bool IncludeExpired { get; set; }

        /// <summary>
        /// Validate continuous query state.
        /// </summary>
        internal void Validate()
        {
            if (Listener == null)
                throw new ArgumentException("Listener cannot be null.");
        }
    }
}
